File: static/CWE-502/python/patched.py
# Remediation for CWE-502: Deserialization of Untrusted Data
# Fix: Parse untrusted input as JSON and use yaml.safe_load instead of pickle/yaml.Loader.

import json
import yaml
from flask import Flask, request

app = Flask(__name__)


@app.route("/load", methods=["POST"])
def load():
    data = request.get_data()
    obj = json.loads(data)
    config = yaml.safe_load(request.args.get("cfg"))
    return str(obj) + str(config)

File: static/CWE-502/javascript/patched.js
// Remediation for CWE-502: Deserialization of Untrusted Data
// Fix: Parse the untrusted input with JSON.parse instead of node-serialize, which cannot execute embedded code.

const express = require('express');

const app = express();
app.use(express.json());

app.post('/load', (req, res) => {
  const data = req.body.data;
  let obj;
  try {
    obj = typeof data === 'string' ? JSON.parse(data) : data;
  } catch (err) {
    return res.status(400).json({ error: 'invalid data' });
  }
  res.json({ loaded: obj });
});

app.listen(3000);
